File: lclang/src/main/java/lclang/statements/expressions/ReturnExpression.java
package lclang.statements.expressions;

import lclang.Caller;
import lclang.LCBaseExecutor;
import lclang.Link;
import lclang.exceptions.LCLangRuntimeException;
import lclang.libs.lang.classes.VoidClass;

public class ReturnExpression extends Expression {
    public final Expression expression;

    public ReturnExpression(Expression expression) {
        super(0);
        this.expression = expression;
    }

    @Override
    public Link visit(Caller prevCaller, LCBaseExecutor visitor) throws LCLangRuntimeException {
        Link value = VoidClass.value;
        if(expression!=null)
            value = expression.visit(prevCaller, visitor);

        value.state = Link.State.RETURN;
        return value;
    }
}

File: lclang/src/main/kotlin/lclang/statements/ReturnExpression.kt
package lclang.statements

import lclang.Caller
import lclang.LCBaseExecutor
import lclang.VOID_VALUE
import lclang.Value

class ReturnExpression(
    val expression: Expression?,
    line: Int,
    column: Int
): Expression(line, column) {
    override fun visit(prevCaller: Caller, visitor: LCBaseExecutor): Value {
        val caller = getCaller(prevCaller)
        var value = VOID_VALUE
        if(expression!=null)
            value = expression.visit(caller, visitor)

        value.state = Value.State.RETURN
        return value
    }
}